targets/psa_switch/externs/psa_internet_checksum.cpp

```diff
@@ -0,0 +1,134 @@
+/* Copyright 2021 SYRMIA LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/*
+ * Dusan Krdzic (dusan.krdzic@syrmia.com)
+ *
+ */
+
+
+#include "psa_internet_checksum.h"
+#include <bm/bm_sim/_assert.h>
+
+
+namespace {
+
+uint16_t 
+ones_complement_sum(uint16_t x, uint16_t y) {
+    uint32_t ret = (uint32_t) x + (uint32_t) y; 
+    if (ret >= 0x10000) 
+        ret++;
+
+    return ret;
+}
+
+}  // namespace
+
+namespace bm {
+
+namespace psa {
+
+void
+PSA_InternetChecksum::init() {
+    clear();
+}
+
+void
+PSA_InternetChecksum::get(Field &dst) const {
+    dst.set(static_cast<uint16_t>(~sum));
+}
+
+void
+PSA_InternetChecksum::get_state(Field &dst) const {
+    dst.set(sum);
+}
+
+void
+PSA_InternetChecksum::get_verify(Field &dst, Field &equOp) const {
+    dst.set(equOp.get<uint16_t>() == static_cast<uint16_t>(~sum));
+}
+
+void
+PSA_InternetChecksum::set_state(const Data &src) {
+    sum = src.get<uint16_t>();
+}
+
+void
+PSA_InternetChecksum::clear() {
+    sum = 0;
+}
+
+void
+PSA_InternetChecksum::add(const std::vector<Field> fields) {
```

Comment:
it feels like this could be `const std::vector<Field> &fields` (same for `subtract`). I don't remember the implementation enough to know if it will make a difference after compiler optimization.

Reply:
The implementation for `field_list` is similar as the implementation for `parameters_vector`. The only difference is that `parameters_vector` supports `std::vector<Data>` as parameter in corresponding method, and `field_list` supports `std::vector<Field>`. Both of them use local vector `vec` in converting method `to` and it won't be correct to return reference to local variable.

`template <> inline
const std::vector<Data>
ActionParam::to<const std::vector<Data>>(ActionEngineState *state) const {
  _BM_ASSERT(tag == ActionParam::PARAMS_VECTOR && "not a params vector");
  std::vector<Data> vec;

  for (auto i = params_vector.start ; i < params_vector.end ; i++) {
    // re-use previously-defined cast method; note that we use to<const Data &>
    // and not to<const Data>, as it does not exists
    // if something in the parameters_vector cannot be cast to "const Data &",
    // the code will assert
    vec.push_back(state->parameters_vector[i].to<const Data &>(state));
  }

  return vec;
}

template <> inline
const std::vector<Field>
ActionParam::to<const std::vector<Field>>(ActionEngineState *state) const {
  _BM_ASSERT(tag == ActionParam::FIELD_LIST && "not a field list");
  std::vector<Field> vec;

  for (auto i = field_list.start ; i < field_list.end ; i++) {
    // re-use previously-defined cast method; note that we use to<const Field &>
    // and not to<const Field>, as it does not exists
    // if something in the field list cannot be cast to "const Field &",
    // the code will assert
    vec.push_back(state->field_list[i].to<const Field &>(state));
  }

  return vec;
}`